Add docs for migrating jobs to EKS cluster (#29908)
kubernetes/test-infra: docs/eks-jobs-migration.md

# Migrating Kubernetes Jobs To The EKS Prow Build Cluster

In an ongoing effort to migrate to community-owned resources, SIG K8S Infra and
SIG Testing are working to complete the migration of jobs from the Google-owned
internal GKE cluster to community-owned clusters.

Most of jobs in the Prow Default Build Cluster should attempt to migrate to
`cluster: eks-prow-build-cluster`. For criteria and requirements, please
pay attention to this document.

## What is EKS Prow Build Cluster?

EKS Prow Build Cluster (`eks-prow-build-cluster`) is an AWS EKS based cluster
owned by community that's used for running ProwJobs. It's supposed to be
similar to our GCP/GKE based clusters, but there are some significant
differences such as:

> - Operating system on worker nodes where we run jobs is Ubuntu 22.04

Reviewer: Actually, it's 20.04 (EKS optimised)
<img width="507" alt="Screenshot 2023-06-30 at 09 12 39" src="https://github.com/kubernetes/test-infra/assets/9121459/bcde2eb7-0dd5-4a0c-a835-4e558e14fec1">

The following jobs **MUST NOT** be migrated at this time:

- Jobs using GCP resources (e.g. E2E tests, promotion jobs, etc.)
- Jobs that are running in trusted clusters (e.g. `test-infra-trusted` and
  `k8s-infra-prow-build-trusted`)

Jobs that are already running in community-owned clusters (e.g. 
`k8s-infra-prow-build`) can be migrated as well, but it's not required or
mandated.

## How To Migrate Jobs?

Fork and check out the [kubernetes/test-infra][k-test-infra] repository,
then follow the steps below:

[k-test-infra]: https://github.com/kubernetes/test-infra

- Find a job that you want to migrate
  - You can check out the following ["Prow Results"][prow-results-default] link
    for recent jobs that are running in the `default` cluster
  - For explanation reasons, let's say that you picked up a job called
    `pull-jobset-test-integration-main`
- Edit the file that `pull-jobset-test-integration-main` is defined in. All
  jobs are defined in the [kubernetes/test-infra][k-test-infra] repository. You
  can use GitHub or our [Code Search tool][cs-test-infra] to find the file
  where this job is defined (search by job name).

> - Look for a `cluster:` key in the job definition. If there isn't one or it's

Reviewer: It might be helpful to point out that it's part of spec for cases where it's not defined.
```suggestion
- Look for a `spec.cluster` key in the job definition.
```

  set to `default`, then the job runs in the default cluster. Add (or replace)
  the following `cluster:` key: `cluster: eks-prow-build-cluster`.
  - **IMPORTANT: if you see any entries under `label` that says `gce` skip this
    job and get back to it the next time as this is not ready to be moved yet.**
  - **IMPORTANT: if you see any entries under `label` that might indicate that
    a job is using some external resource (e.g. some cloud platform), you need
    to check if the Community Ownership criteria is satisfied for that resource
    (see the Community Ownership section for more details)**
  - **IMPORTANT: jobs running in community-owned clusters must have resource
    requests and limits specified for cluster-autoscaler to work properly.
    If that's not the case for your job, please see the next section for some
    details about determining correct resource requests and limits.**
  - Here's an [example of a job][example-eks-job] as a reference (pay attention
    to the `cluster:` key)
- Save the file, commit the change, create a branch and file a PR
- Once the PR is merged, follow the guidelines in the Post Migration Period

>   section of this document to ensure that the job remains stable

Reviewer: I would add here to check job definition for images referencing docker.io. Last week we had a problem with rate limiting. It's not strictly related to EKS migration but we still should encourage people to replace them with ECR mirrors from [the official gallery](https://gallery.ecr.aws/).

Author: @dims mentioned that we might want to first sync with AWS folks before doing this on large scale, so I'd leave it from recommendations until then.

If you have any trouble, please leave a note in the following
[GitHub issue][test-infra-gh-issue] or come to
[#sig-k8s-infra][slack-k8s-infra] or [#sig-testing][slack-sig-testing] Slack

> channel to ask for help.

Reviewer: This block is duplicated. I'm ok with that, but we could also redirect to the bottom part of this document.

[prow-results-default]: https://prow.k8s.io/?cluster=default
[cs-test-infra]: https://cs.k8s.io/?q=job-name-here&i=nope&files=config%2F&excludeFiles=&repos=kubernetes/test-infra
[example-eks-job]: https://github.com/kubernetes/test-infra/blob/1d219efcca8a254aaca2c34570db0a56a05f5770/config/jobs/kubernetes/cloud-provider-aws/cloud-provider-aws-config.yaml#L3C1-L31
[test-infra-gh-issue]: https://github.com/kubernetes/test-infra/issues/29722
[slack-k8s-infra]: https://kubernetes.slack.com/archives/CCK68P2Q2
[slack-sig-testing]: https://kubernetes.slack.com/archives/C09QZ4DQB

### Determining Resource Requests and Limits

Jobs running in community-owned clusters must have resource requests and 
limits specified for cluster-autoscaler to work properly. However, determining
this is not always easy and incorrect requests and limits can cause the job

> to start flaking or failing.

Reviewer: Or even affect other jobs running in the cluster in case it causes CPU throttling.

Author: AFAIK, throttling only applies to a specific cgroupns and each pod is supposed to have its own namespace. That being said, hopefully, other jobs shouldn't be affected by throttling.

    if the job has a testgrid dashboard
  - The job duration and success/flakiness rate should be same or similar as in
    the old cluster. If you notice significant bumps, try adjusting resource
    requests and limits. If that doesn't help, reach out to us so that we
    can investigate
  - Note that some jobs are flaky by their nature, i.e. they were flaking in
    the default cluster too. This is not going to be fixed by moving job to
    the new cluster, but we shouldn't see flakiness rate getting worse
- Watch job's actual resource consumption and adjust requests and limits as
  needed
  - We have a [Grafana dashboard][monitoring-eks] that can show you actual
    resource CPU and memory usage for job. If you notice that CPU gets
    throttled too often, try increasing number of allowed CPUs. Similar for
    memory, if you see memory usage too close too limits, try increasing it
    a little bit

[prow-results]: https://prow.k8s.io
[testgrid]: http://testgrid.k8s.io
[monitoring-eks]: https://monitoring-eks.prow.k8s.io/d/53g2x7OZz/jobs?orgId=1&refresh=30s&var-org=kubernetes&var-repo=kubernetes&var-job=All

## Known Issues

> - Go doesn't respect cgroups, i.e. CPU limits are not going to be respected by

Reviewer: I might exaggerate here, but using full name will be less ambiguous. 
```suggestion
- Golang/Go doesn't respect cgroups, i.e. CPU limits are not going to be respected by
```

  Go applications
  - This means that a Go application will try to use all available CPUs on

>     node, even though it's limited, e.g. to 2

Reviewer: ```suggestion
    node, even though it's limited by k8s resource limit, e.g. to 2
```

  - This can cause massive throttling and performance-sensitive tasks and tests
    can be hit by this. Nodes in the new clusters are much larger
    (we had 8 vCPUs in the old cluster per node, while we have 16 vCPUs in the
    new cluster per node), so it can be easier to get affected by this issue
  - In general, number of tests affected by this should be **very low**.
    However, if you think you're affected by this, you can try to mitigate the
    issue by setting `GOMAXPROCS` environment variable for that job to the
    value of `cpu` limit. There are also ways to automatically determine
    `GOMAXPROCS`, such as [`automaxprocs`][automaxprocs]

> - Kernel configs are not available inside the job's test pod so kubeadm can

Reviewer: `can` or `cannot`?
```suggestion
- Kernel configs are not available inside the job's test pod so kubeadm cannot
```

Author: Replaced with `might`.
